### ISSUE TITLE
Idiomatic way of keeping input's value in sync with state?

### ISSUE THREAD

**Issue by maximilianschmitt** [marko-js/marko#640]
Hi!

I've been developing with React for the past 2 years and using Marko really feels nice and fresh. :) I love how fast and lightweight it feels

But one thing I couldn't figure out and found no examples for, is keeping the `value` of an `input` in sync with the component's `state`. With React, you do `<input value={state.value} onChange={this.onChange} />` where `onChange` calls `setState({ value: e.target.value })` and it just works.

Intuitively I thought the correct way to go about this in Marko would be to do something like this:

```
class {
  onCreate() {
    this.state = { value: 'Hello' }
  }

  handleInput(e) {
    this.state.value = e.target.value
  }
}

input value=state.value on-input('handleKeydown')
-- ${state.value}
```

`keyDown` is being called, however the component is not rerendering if I type in the input. The text next to it (`-- ${state.value}`) always says `Hello`. Also, the `input` only seems to pick up every second character I type in. If I add a `setStateDirty('value')` to my `handleInput`-function, the input's value always stays at `Hello`. I can tell (by logging) that in this case the component is rerendering (with no visible result though).

Is my approach fundamentally wrong? How do you normally keep state and input components in sync?

Thanks for your help!

Best,
Max

**Comment by mindeavor** [comment]
Your approach looks good to me. However, you're trying to do root-level text, which may mess things up. Try wrapping your `input` and text within a `<div>`.

**Comment by maximilianschmitt** [comment]
@mindeavor ah, thank you! :)

**Comment by patrick-steele-idem** [comment]
Just wanted to mention that Marko has a [known limitation](https://github.com/marko-js/marko/issues/589) that does not allow dynamic top-level HTML elements or text nodes. Once we fix #589 then the wrapping `<div>` would no longer be needed. We have some ideas on how to fix the issue, but it's a little tricky.
